### PR TITLE
[fix](expr) Fix BE core dump while common expr filter delete condition column

### PR DESCRIPTION
**Pull request by xinyiZzz**
## Proposed changes

pred column also needs to be filtered by expr, exclude delete condition column, delete condition column not need to be filtered, query engine does not need it, after _output_column_by_sel_idx, delete condition materialize column will be erase at the end of the block.
Eg:
 `delete from table where a = 10;`
 `select b from table;`
a column only effective in segment iterator, the block from query engine only contain the b column, so no need to filter a column by expr.

```
*** SIGSEGV address not mapped to object (@0x0) received by PID 145953 (TID 146447 OR 0x7ffecbf78700) from PID 0; stack trace: ***
 0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /root/src/doris-2.0/be/src/common/signal_handler.h:417
 1# os::Linux::chained_handler(int, siginfo*, void*) in /home/data1/apache-doris-2.0.3-rc01-bin-x64/java8/jre/lib/amd64/server/libjvm.so
 2# JVM_handle_linux_signal in /home/data1/apache-doris-2.0.3-rc01-bin-x64/java8/jre/lib/amd64/server/libjvm.so
 3# signalHandler(int, siginfo*, void*) in /home/data1/apache-doris-2.0.3-rc01-bin-x64/java8/jre/lib/amd64/server/libjvm.so
 4# 0x00007FFFF7621400 in /lib64/libc.so.6
 5# doris::vectorized::Block::filter_block_internal(doris::vectorized::Block*, std::vector<unsigned int, std::allocator<unsigned int> > const&, doris::vectorized::PODArray<unsigned char, 4096ul, Allocator<false, false, false>, 15ul, 16ul> const&) at /root/src/doris-2.0/be/src/vec/core/block.cpp:738
 6# doris::vectorized::VExprContext::execute_conjuncts_and_filter_block(std::vector<std::shared_ptr<doris::vectorized::VExprContext>, std::allocator<std::shared_ptr<doris::vectorized::VExprContext> > > const&, doris::vectorized::Block*, std::vector<unsigned int, std::allocator<unsigned int> >&, int, doris::vectorized::PODArray<unsigned char, 4096ul, Allocator<false, false, false>, 15ul, 16ul>&) at /root/src/doris-2.0/be/src/vec/exprs/vexpr_context.cpp:284
 7# doris::segment_v2::SegmentIterator::_execute_common_expr(unsigned short*, unsigned short&, doris::vectorized::Block*) at /root/src/doris-2.0/be/src/olap/rowset/segment_v2/segment_iterator.cpp:2127
 8# doris::segment_v2::SegmentIterator::_next_batch_internal(doris::vectorized::Block*) in /home/data1/apache-doris-2.0.3-rc01-bin-x64/be/lib/doris_be
 9# doris::segment_v2::SegmentIterator::next_batch(doris::vectorized::Block*) at /root/src/doris-2.0/be/src/olap/rowset/segment_v2/segment_iterator.cpp:1876
10# doris::BetaRowsetReader::next_block(doris::vectorized::Block*) at /root/src/doris-2.0/be/src/olap/rowset/beta_rowset_reader.cpp:298
11# doris::vectorized::VCollectIterator::Level0Iterator::refresh_current_row() at /root/src/doris-2.0/be/src/vec/olap/vcollect_iterator.cpp:511
12# doris::vectorized::VCollectIterator::Level0Iterator::ensure_first_row_ref() at /root/src/doris-2.0/be/src/vec/olap/vcollect_iterator.cpp:490
13# doris::vectorized::VCollectIterator::Level1Iterator::ensure_first_row_ref() at /root/src/doris-2.0/be/src/vec/olap/vcollect_iterator.cpp:689
14# doris::vectorized::VCollectIterator::build_heap(std::vector<std::shared_ptr<doris::RowsetReader>, std::allocator<std::shared_ptr<doris::RowsetReader> > >&) at /root/src/doris-2.0/be/src/vec/olap/vcollect_iterator.cpp:184
15# doris::vectorized::BlockReader::_init_collect_iter(doris::TabletReader::ReaderParams const&) at /root/src/doris-2.0/be/src/vec/olap/block_reader.cpp:147
16# doris::vectorized::BlockReader::init(doris::TabletReader::ReaderParams const&) at /root/src/doris-2.0/be/src/vec/olap/block_reader.cpp:226
17# doris::vectorized::NewOlapScanner::open(doris::RuntimeState*) at /root/src/doris-2.0/be/src/vec/exec/scan/new_olap_scanner.cpp:224
18# doris::vectorized::ScannerScheduler::_scanner_scan(doris::vectorized::ScannerScheduler*, doris::vectorized::ScannerContext*, std::shared_ptr<doris::vectorized::VScanner>) at /root/src/doris-2.0/be/src/vec/exec/scan/scanner_scheduler.cpp:345
19# std::_Function_handler<void (), doris::vectorized::ScannerScheduler::_schedule_scanners(doris::vectorized::ScannerContext*)::$_1::operator()() const::{lambda()#3}>::_M_invoke(std::_Any_data const&) at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:291
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/vec/core/block.cpp [hunk 1]
```diff
@@ -730,14 +730,40 @@ void Block::update_hash(SipHash& hash) const {
 
 void Block::filter_block_internal(Block* block, const std::vector<uint32_t>& columns_to_filter,
                                   const IColumn::Filter& filter) {
+    auto block_column_not_exist = [&columns_to_filter, block](auto& col) -> void {
+        std::ostringstream ss;
+        for (const auto& i : columns_to_filter) {
+            ss << i << "-";
```
[the remark this answers is not in the document]
Comment (xinyiZzz) [reply be/src/vec/core/block.cpp]: 这个只会在 column 是 nullptr 的时候执行，不会影响正常执行

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

File: regression-test/suites/correctness/test_pushdown_common_expr.groovy [hunk 2]
```diff
@@ -150,4 +150,62 @@ suite("test_pushdown_common_expr") {
...
+        `c5` varchar(100) NULL COMMENT "",
+        `c6` varchar(100) NULL COMMENT "",
+        `c7` varchar(100) NULL COMMENT ""
+    ) ENGINE=OLAP
+    DUPLICATE KEY(`c1`)
+    COMMENT 'OLAP'
+    DISTRIBUTED BY HASH(`c1`) BUCKETS 1
+    PROPERTIES (
+        "replication_allocation" = "tag.location.default: 1",
+        "disable_auto_compaction" = "false"
```
Comment (yiguolei) [inline regression-test/suites/correctness/test_pushdown_common_expr.groovy:170]: true

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 3]
```diff
@@ -1598,8 +1601,9 @@ Status SegmentIterator::_vec_init_lazy_materialization() {
         if (!_common_expr_columns.empty()) {
             _is_need_expr_eval = true;
             for (auto cid : _schema->column_ids()) {
-                // pred column also needs to be filtered by expr
-                if (_is_common_expr_column[cid] || _is_pred_column[cid]) {
+                // pred column also needs to be filtered by expr, exclude additional deleted column,
```
[the remark this answers is not in the document]
Comment (xinyiZzz) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp]: done

File: regression-test/suites/correctness/test_pushdown_common_expr.groovy (hunk 2 above)
Comment (xinyiZzz) [reply regression-test/suites/correctness/test_pushdown_common_expr.groovy:170]: fixed

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by zhiqiang-hhhh — APPROVED** [review]
LGTM

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**